notebooks/MAST/astroqueryMAST/beginner_program_id_search/companion_script.py

```diff
@@ -0,0 +1,48 @@
+import sys
+from astroquery.mast import Observations
+
+def download_by_jwst_pid (pid, curl_flag):
+    """
+    This function retrieves all of the level 2 science products for a single program ID.
+    Inputs:
+    pid [int] – program ID number
+    curl_flag [Boolean] – If set to true, function returns a curl script.
+    """
+
+    # Login, search for observations
+    Observations.login(store_token=True)
+    obs_list = Observations.query_criteria(proposal_id=pid)
+    data_products = Observations.get_product_list(obs_list)
+    filtered_prod = Observations.filter_products(data_products, calib_level=[2], productType='SCIENCE')
+
+    # Filesize check
+    total = 0
+    for fsize in filtered_prod['size']:
+        total = total+fsize
+    # Use curl for large downloads.
+    if total/10**9 > 1 and curl_flag==False:
+        curl_flag = True
+        print("\nDownload too large. Switching to curl script.")
+    # User must confirm download size
+    prompt = "Total download size is: " + '{:.2f}'.format(total/10**9) + ' GB. Proceed? (y/n)\n'
+    resp = input(prompt)
+    if resp != 'y':
+        quit()
+
+    # Check for curl flag
+    if curl_flag:
+        path = Observations.download_products(filtered_prod, curl_flag=True)
+        return path
+    else:
+        manifest = Observations.download_products(filtered_prod)
+        return manifest
+
+def main():
+    return download_by_jwst_pid(a, b)
+
+if __name__ == "__main__":
```

Comment:
If we want to be a little fancier with command line arguments, we should use `[argparse](https://docs.python.org/3/library/argparse.html)`.  I can provide some help/examples, but if would make this part easier to understand, provide output for "python companion_script.py --help" with documentation that will say what the command line arguments are, etc.

Reply:
If so, then the block of code in __main__ would look more like:

if __name__ == "__main__":
  (setup argparse)
  args = parser.parse_args()
  download_by_jwst_pid(args.pid, args.curl_flag)

I like to do my argparse setup in its own function, e.g., make a "setup_args()", so then it would be:

if __name__ == "__main__":
  args = setup_args()
  download_by_jwst_pid(args.pid, args.curl_flag)

and:

def setup_args():
  parser = argparse.ArgumentParser(description='Top-level description.')
  parser.add_argument(something for pid)
  parser.add_argument(something for curl_flag)
  return parser.parse_args()

Reply:
and in this case, you won't even need to define a main(): at all.